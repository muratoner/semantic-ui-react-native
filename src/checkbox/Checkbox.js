import React, { useMemo } from 'react';
import {
	LayoutAnimation,
	StyleSheet,
	Text,
	TouchableOpacity,
	View
} from 'react-native';
import { CheckboxProps } from '../';
import Icon from '../icons/Icon';

export default (props: CheckboxProps) => {
	const {
		style,
		icon,
		iconType = 'FontAwesome',
		containerStyle,
		iconStyle,
		titleStyle,
		checked,
		onChanged,
		type = 'circle',
		checkedColor,
		unCheckedColor,
		disable,
		title
	} = props;

	const Slider = useMemo(() => {
		if (type == 'slider') {
			return (
				<View
					style={[
						styles.slider,
						{
							backgroundColor: checked ? '#191919' : 'rgba(0,0,0,.1)'
						}
					]}
				>
					<View
						style={{
							borderRadius: 50,
							shadowColor: 'rgba(0,0,0,.5)',
							shadowOpacity: 1,
							shadowRadius: 1,
							shadowOffset: { width: 0, height: 0 },
							width: 20,
							height: 20,
							backgroundColor: 'white',
							position: 'absolute',
							top: -8,
							left: !checked ? 0 : undefined,
							right: checked ? 0 : undefined
						}}
					></View>
				</View>
			);
		}
	}, [type, checked]);

	const Toggle = useMemo(() => {
		if (type == 'toggle') {
			return (
				<View
					style={[
						styles.toggle,
						{
							backgroundColor: checked ? '#2185d0' : 'rgba(0,0,0,.1)'
						}
					]}
				>
					<View
						style={[
							styles.toggleCircle,
							{
								left: !checked ? 1 : undefined,
								right: checked ? 1 : undefined
							}
						]}
					></View>
				</View>
			);
		}
	}, [type, checked]);

	const IconComponent = useMemo(
		() => () => {
			if (typeof icon == 'object') {
				return icon;
			}
			let name = '';
			let iconType = '';
			switch (type) {
			case 'circle':
				name = checked ? 'check-circle-outline' : 'circle-outline';
				iconType = 'MaterialCommunityIcons';
				break;
			case 'heart':
				name = checked ? 'heart-sharp' : 'heart-outline';
				iconType = 'Ionicons';
				break;
			case 'circle-overflow':
				name = checked ? 'check-circle' : 'circle';
				iconType = 'Feather';
				break;
			case 'volume':
				name = checked ? 'ios-volume-high' : 'ios-volume-mute';
				iconType = 'Ionicons';
				break;
			case 'square-overflow':
				name = checked ? 'check-square' : 'square';
				iconType = 'Feather';
				break;
			case 'square':
				name = checked ? 'checksquareo' : 'square';
				iconType = checked ? 'AntDesign' : 'Feather';
				break;
			case 'notification':
				name = checked
					? 'ios-notifications-circle-outline'
					: 'ios-notifications-off-circle-outline';
				iconType = 'Ionicons';
				break;
			case 'lock':
				name = checked ? 'lock-closed' : 'lock-open';
				iconType = 'Ionicons';
				break;
			case 'flag':
				name = checked ? 'flag-checkered' : 'flag-o';
				iconType = 'FontAwesome';
				break;
			case 'radio':
				name = checked ? 'radio-button-on' : 'radio-button-off';
				iconType = 'Ionicons';
				break;
			case 'slider':
				return Slider;
			case 'toggle':
				return Toggle;
			}
			return (
				<Icon
					name={name}
					type={iconType}
					color={(checked ? checkedColor : unCheckedColor) || '#191919'}
					style={StyleSheet.flatten([style, styles.icon(type), iconStyle])}
				/>
			);
		},
		[type, iconStyle, checked, style, checkedColor, unCheckedColor, disable]
	);

	return (
		<TouchableOpacity
			onPress={
				!disable
					? () => {
						LayoutAnimation.easeInEaseOut();
						onChanged?.(!checked);
					  }
					: undefined
			}
			style={[styles.container, containerStyle, { opacity: disable ? 0.7 : 1 }]}
		>
			<IconComponent />
			{title && <Text style={[styles.title, titleStyle]}>{title}</Text>}
		</TouchableOpacity>
	);
};

const styles = StyleSheet.create({
	container: {
		flexDirection: 'row',
		alignItems: 'center'
	},
	title: {
		marginLeft: 10
	},
	slider: {
		height: 3,
		width: 50,
		borderRadius: 10
	},
	toggle: {
		height: 20,
		width: 50,
		borderRadius: 20 * 0.5
	},
	toggleCircle: {
		borderRadius: 50,
		shadowColor: 'rgba(0,0,0,.5)',
		shadowOpacity: 1,
		shadowRadius: 1,
		shadowOffset: { width: 0, height: 0 },
		width: 18,
		height: 18,
		backgroundColor: 'white',
		position: 'absolute',
		top: 1
	},
	icon: (type) => ({
		padding: type == 'flat' ? 2 : null,
		borderRadius: 5,
		fontSize: 24,
		borderColor: '#d4d4d5',
		borderWidth: type == 'flat' ? 1 : undefined,
		alignSelf: 'center'
	})
});
