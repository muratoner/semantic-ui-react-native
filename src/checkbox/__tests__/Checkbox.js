import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import Checkbox from '../Checkbox';

describe('Checkbox Component', () => {
	it('should render without issues', () => {
		const component = shallow(<Checkbox />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render slider', () => {
		const component = shallow(<Checkbox type="slider" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render circle', () => {
		const component = shallow(<Checkbox type="circle" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render circle-overflow', () => {
		const component = shallow(<Checkbox type="circle-overflow" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render flag', () => {
		const component = shallow(<Checkbox type="flag" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render heart', () => {
		const component = shallow(<Checkbox type="heart" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render lock', () => {
		const component = shallow(<Checkbox type="lock" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render notification', () => {
		const component = shallow(<Checkbox type="notification" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render radio', () => {
		const component = shallow(<Checkbox type="radio" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render square', () => {
		const component = shallow(<Checkbox type="square" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render square-overflow', () => {
		const component = shallow(<Checkbox type="square-overflow" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render toggle', () => {
		const component = shallow(<Checkbox type="toggle" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render volume', () => {
		const component = shallow(<Checkbox type="volume" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render volume', () => {
		const component = shallow(<Checkbox type="volume" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render title', () => {
		const component = shallow(<Checkbox type="volume" title="Volume toggle" />);
		expect(component).toHaveLength(1);

		expect(toJson(component)).toMatchSnapshot();
	});
});
