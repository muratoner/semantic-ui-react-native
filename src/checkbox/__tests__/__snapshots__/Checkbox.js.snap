// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component should render circle 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render circle-overflow 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render flag 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render heart 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render lock 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render notification 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render radio 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render slider 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render square 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render square-overflow 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render title 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
  <Text
    style={
      Array [
        Object {
          "marginLeft": 10,
        },
        undefined,
      ]
    }
  >
    Volume toggle
  </Text>
</ForwardRef>
`;

exports[`Checkbox Component should render toggle 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render volume 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render volume 2`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;

exports[`Checkbox Component should render without issues 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
      Object {
        "opacity": 1,
      },
    ]
  }
>
  <Component />
</ForwardRef>
`;
